[REST Compatible API] Route refactoring

joegallo opened the pull request:
Related to #51816

Makes `Route`s  `RestApiVersion` -aware (and `RestHandler`s `RestApiVersion` -agnostic). Refactors how `Route`s are constructed in the case of deprecation or replacement of routes.

-----------

### Some examples of how this affects Rest Actions

Very normal existing rest action, no change on this PR:

```
public class RestMainAction extends BaseRestHandler {

    @Override
    public List<Route> routes() {
        return List.of(
            new Route(GET, "/"),
            new Route(HEAD, "/"));
    }

[...]
```

A rest action with a deprecated route, before:

```
public class RestPostDataAction extends BaseRestHandler {

    @Override
    public List<Route> routes() {
        return Collections.emptyList();
    }

    @Override
    public List<DeprecatedRoute> deprecatedRoutes() {
        return Collections.singletonList(
            new DeprecatedRoute(POST, MachineLearning.BASE_PATH + "anomaly_detectors/{" + Job.ID.getPreferredName() + "}/_data",
            "Posting data directly to anomaly detection jobs is deprecated, " +
                "in a future major version it will be compulsory to use a datafeed"));
    }

[...]
```

And after:

```
public class RestPostDataAction extends BaseRestHandler {

     @Override
     public List<Route> routes() {
        return List.of(
            Route.builder(POST, BASE_PATH + "anomaly_detectors/{" + Job.ID.getPreferredName() + "}/_data")
                .deprecated("Posting data directly to anomaly detection jobs is deprecated, " +
                    "in a future major version it will be compulsory to use a datafeed", RestApiVersion.V_8
                ).build()
        );
    }

[...]
```



A rest action with some replaced routes, before:

```
public class RestGetUsersAction extends SecurityBaseRestHandler {

    @Override
    public List<Route> routes() {
        return Collections.emptyList();
    }

    @Override
    public List<ReplacedRoute> replacedRoutes() {
        // TODO: remove deprecated endpoint in 8.0.0
        return List.of(
            new ReplacedRoute(GET, "/_security/user/", GET, "/_xpack/security/user/"),
            new ReplacedRoute(GET, "/_security/user/{username}", GET, "/_xpack/security/user/{username}")
        );
    }

[...]
```

And after:

```
public class RestGetUsersAction extends SecurityBaseRestHandler {

    @Override
    public List<Route> routes() {
        return List.of(
           Route.builder(GET, "/_security/user/")
               .replaces(GET, "/_xpack/security/user/", RestApiVersion.V_7).build(),
           Route.builder(GET, "/_security/user/{username}")
               .replaces(GET, "/_xpack/security/user/{username}", RestApiVersion.V_7).build()
        );
    }

[...]
```

joegallo committed:
Simplify RestHandler to pass all routes through routes()

joegallo committed:
Whitespace

joegallo committed:
Replaced class hierarchy with fluent interface

joegallo commented:
Crucially, these sorts of methods fall out:

```
         @Override
         public List<DeprecatedRoute> deprecatedRoutes() {
             return [...];
         }

         @Override
         public List<ReplacedRoute> replacedRoutes() {
             return [...];
         }
```

And now there is only `routes()`.

jaymode commented on x-pack/plugin/deprecation/qa/rest/src/main/java/org/elasticsearch/xpack/deprecation/TestDeprecationHeaderRestAction.java:
```diff
@@ -77,19 +75,14 @@ public TestDeprecationHeaderRestAction(Settings settings) {
     @Override
     public List<Route> routes() {
-        return Collections.emptyList();
+        return List.of(new Route(GET, "/_test_cluster/deprecated_settings").deprecated(DEPRECATED_ENDPOINT));
```
I prefer a builder pattern over a method that returns a new instance. Just reading this line of code makes me think that the `Route` object is mutable when it is not and instead the method returns a new instance. With a builder pattern, I find the code has more clarity at the expense of conciseness.

```java
Route.builder(GET, "/_test_cluster/deprecated_settings").setDeprecationMessage(DEPRECATED_ENDPOINT).build();
```

Static methods for construction similar to the JDK style `of` syntax may provide an alternative that is middle ground for the two?

```java
// normal
Route.of(GET, "/_test_cluster/deprecated_settings");
// a deprecated route
Route.deprecated(GET, "/_test_cluster/deprecated_settings", DEPRECATED_ENDPOINT);
// a replaced route
Route.replaced(GET, "/_test_cluster/deprecated_settings", GET, "/_test_cluster/deprecated_settings2");
// a compatible route - V7 only
Route.compatibleWith(GET, "/_test_cluster/deprecated_settings", CompatibleVersion.V7);
// a compatible route range - works with both 7 and 8
Route.compatibleWith(GET, "/_test_cluster/deprecated_settings", new CompatibleRange(CompatibleVersion.V7, CompatibleVersion.V8));
```

Maybe this makes sense or sparks another idea that's better :)

joegallo replied:
+1, I'll put this in my pipe and smoke it. Thanks for taking a look!

joegallo replied:
I ended up something pretty near what you'd suggested, but a little different. It's got a `Route.of` for the trivial case and a builder for some of the more interesting stuff. Nothing set in stone of course, as you'll be able to see from my string of "ooh, wait, what if we did it like *this*!" commits. 😄

joegallo committed:
JDK-style of syntax

joegallo committed:
You must give a version argument for replaces()

joegallo committed:
You have to specify the version you deprecated something in

joegallo committed:
Rework to a builder

joegallo committed:
Twiddle imports to conserve precious characters

joegallo committed:
Rename this method

joegallo committed:
Merge remote-tracking branch 'upstream/master' into route-refactoring

pgomulka commented:
I like the builder approach. Not sure we need the `of` method. The name does not say much more than the `new Route`
And since we have the `new Route.Builder` as well it would fit nicely with just `new Route` (also the refactoring would be smaller thanks to this).
But either way it looks good to me

joegallo commented:
> Not sure we need the of method.

Yeah, if this were a greenfield project, I might be inclined to go with `Route.of` but in this existing codebase I agree that it may not be worth the churn. -- @jaymode what do you think?

joegallo committed:
Route.of back to new Route

jaymode commented:
I agree that `Route.of` can be dropped

joegallo committed:
Undo many now-extraneous whitespace changes

joegallo committed:
Merge remote-tracking branch 'upstream/master' into route-refactoring

joegallo committed:
Twiddle imports

joegallo committed:
Rework MethodHandlers a little bit

joegallo committed:
Emphasize the Route as the relevant concept

joegallo committed:
Rename this method

joegallo committed:
Oops, fix an out of date method reference

joegallo committed:
Move version awareness to Route from RestHandler

joegallo committed:
Dispatch to deprecation or compatibility log by version

joegallo committed:
Merge remote-tracking branch 'upstream/master' into route-refactoring

joegallo committed:
The deprecation tests need to be currentVersion

If it's the previous version, then it exercises the rest compatibility
path rather than the vanilla deprecation path.

joegallo committed:
Actually, we're keeping POST _data for 8.x (for now)

This could change later, but based on a conversation with the
machine-learning team, we don't want to commit to dropping this in 8.0
right this moment.

joegallo committed:
Add compatible-with=7 when using the deprecated /transform/ routes

joegallo committed:
Revert "Add compatible-with=7 when using the deprecated /transform/ routes"

This reverts commit b6686d77a06890bf601348a8b3ae9ea31b4b78ec.

joegallo committed:
Temporarily mark these ML APIs as deprecated in 8.x

There are java and docs tests, etc, that expect them to be there, and
it's inconvenient to need to fix that now. Instead, we can temporarily
mark them as deprecated in 8, with the intention of looping back
around to the docs and tests later -- when that's done, then these can
all be marked as deprecated in 7, which will mean they're still
available in 8 via compatible-with=7.

pgomulka commented on server/src/main/java/org/elasticsearch/rest/MethodHandlers.java:
```diff
@@ -20,28 +21,23 @@
-            methodHandlers.computeIfAbsent(method, k -> new HashMap<>())
-                .put(handler.compatibleWithVersion(), handler);
-        }
+        this.methodHandlers = new HashMap<>(1);
```
isn't it common that we have multiple methods per same endpoint?

joegallo replied:
Yes, but the default capacity for a HashMap is 16, and we create about 600 MethodHandlers in the course of starting Elasticsearch. ~450 of those have just one Route in them. ~100 have 2, and the rest have more (a few have as many as five).

It seemed to me like the cost of the HashMap resizing was worth the memory saving, but I ran a little bit of analysis and I'm going to add a commit that bumps the default initial capacity to 4.

joegallo replied:
0b50128

pgomulka commented on server/src/main/java/org/elasticsearch/rest/RestController.java:
```diff
@@ -151,32 +161,42 @@ protected void registerWithDeprecatedHandler(RestRequest.Method method, String p
+                route.getDeprecationMessage());
+        } else {
+            // it's just a normal route
+            registerHandler(route.getMethod(), route.getPath(), RestApiCompatibleVersion.currentVersion(), handler);
```
so what if a handler was not deprecated previously? The deprecation logging was just done within the RestHandler itself (often scenario in types removal)
Shouldn't we get a `RestApiCompatibleVersion` from a Route in this case?

joegallo replied:
Let's catch up on this one over slack or zoom tomorrow.

pgomulka commented on server/src/test/java/org/elasticsearch/rest/RestControllerTests.java:
```diff
@@ -634,20 +637,12 @@ public void testDispatchCompatibleHandler() {
-            public RestApiCompatibleVersion compatibleWithVersion() {
-                return RestApiCompatibleVersion.minimumSupported();
-            }
+        restController.registerHandler(new Route(GET, "/foo"), (request, channel1, client) -> {
```
so this does not indicate that a Route is compatible only. It passes because if you are using compatible api it allows you to access Current Version apis (not compatible) - this is covered in testDispatchCompatibleRequestToNewlyAddedHandler
We could possibly make that test stricker, by sending another fakeRequest for version = current and expecting 404

joegallo replied:
Good catch -- does 65379b80fa22e8af1dc9b54c4fbe7490231addfc cover what you were hoping to see?

pgomulka commented on server/src/test/java/org/elasticsearch/rest/RestHttpResponseHeadersTests.java:
```diff
@@ -49,8 +50,8 @@ public void testUnsupportedMethodResponseHttpHeader() throws Exception {
          * method list, passing in the RandomizedContext's Random instance,
          * before picking out a candidate sublist.
          */
-        List<RestRequest.Method> validHttpMethodArray = new ArrayList<RestRequest.Method>(Arrays.asList(RestRequest.Method.values()));
```
just a style question. Was replacing `RestRequest.Method` with `Method` and an import intentional? I dont' have strong opinion agains this, but using `RestRequest.Method` was more explicit that this lives within RestRequest. I am not sure the context of the `Method` class is that obvious from here.
It would be the same for common usage of `Map.Entry` vs `Entry`
This applies to other places in the PR where an import was for nested class. 
Also it looks like this is probably not documented, but a convention in ES (correct me if my observation is wrong)

joegallo replied:
It was intentional, I think it's a bit prettier in this case, but it's not a huge deal to me and I'll happily back it out if you'd rather I left it alone.

pgomulka replied:
i think it would be better to refactor the nested classes imports separately. This would help to make the PR smaller (we are touching a lot of places with that refactoring, not only related to RestRequest.METHOD)

joegallo replied:
7bb0919 covers this, I twiddled things back.

pgomulka reviewed (commented):
looks great, I left few comments

jakelandis commented on x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/rest/inference/RestGetTrainedModelsStatsAction.java:
```diff
@@ -9,37 +9,31 @@
-                GET, MachineLearning.BASE_PATH + "trained_models/_stats",
-                GET, MachineLearning.BASE_PATH + "inference/_stats"));
+            Route.builder(GET, BASE_PATH + "trained_models/{" + MODEL_ID.getPreferredName() + "}/_stats")
+                .replaces(GET, BASE_PATH + "inference/{" + MODEL_ID.getPreferredName() + "}/_stats", RestApiCompatibleVersion.V_8).build(),
```
What is the backport for this going to look like ?

i.e. planning to back port the RestApiCompatibleVersion class, or adjust the replaces method to only take 1 argument ?

joegallo replied:
[n.b. we've talked about this a bit offline]

We've renamed RestApiCompatibleVersion to RestApiVersion, and it is my intention to backport that class to 7.x, and continue to have the arities of `deprecated` and `replaces` that require it, but not to backport any of the actual rest compatible api backing machinery.

That is, the backport for this stanza is going to look more or less exactly the same as this stanza itself already looks.

jakelandis commented on x-pack/plugin/deprecation/qa/rest/src/main/java/org/elasticsearch/xpack/deprecation/TestDeprecationHeaderRestAction.java:
```diff
@@ -77,19 +75,18 @@ public TestDeprecationHeaderRestAction(Settings settings) {
-        return Collections.emptyList();
+        return List.of(
+            Route.builder(GET, "/_test_cluster/deprecated_settings")
+                .deprecated(DEPRECATED_ENDPOINT, RestApiCompatibleVersion.currentVersion())
```
might want a comment here stating that .currentVersion() is for testing purposes only and actual version should be used in real code. I know i will read the tests for examples of usage and that comment could help to prevent incorrect usage.

joegallo replied:
c178121

jakelandis commented on x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/rest/job/RestPostDataAction.java:
```diff
@@ -27,15 +27,12 @@
+                .deprecated("Posting data directly to anomaly detection jobs is deprecated, " +
+                    "in a future major version it will be compulsory to use a datafeed", RestApiCompatibleVersion.V_8
+                ).build()
+        );
```
Same question about what the back port to 7.x will look like ?

joegallo replied:
Marking this one as resolved, let's carry on the conversation in the other thread (https://github.com/elastic/elasticsearch/pull/69573#discussion_r586460388), though.

jakelandis commented on server/src/main/java/org/elasticsearch/rest/action/admin/cluster/RestCancelTasksAction.java:
```diff
@@ -38,8 +38,10 @@ public String getName() {
     public List<Route> routes() {
-        return List.of(new Route(POST, "/_tasks/_cancel"),
-            new Route(POST, "/_tasks/{task_id}/_cancel"));
+        return List.of(
```
for the backport, you will want to be aware of https://github.com/elastic/elasticsearch/pull/54183 (if you are not already)

joegallo replied:
👍, I've pregamed the backport locally, and that's a little annoying but not a show stopper. Just a bit more tedious manual work.

joegallo committed:
Merge remote-tracking branch 'upstream/master' into route-refactoring

joegallo committed:
Merge remote-tracking branch 'upstream/master' into route-refactoring

joegallo committed:
Switch the MethodHandlers default capacity to 4

We create about 600 of these, ~450 have just a single route, and ~100
of which have two routes. The default capacity of 16 gives is too
high, but with 4 we have enough slots available for the vast majority
of MethodHandlers to hold all their routes without a single resize.

joegallo committed:
Add an explanatory comment

joegallo committed:
Merge remote-tracking branch 'upstream/master' into route-refactoring

joegallo committed:
Future major versions are not expected

joegallo commented on server/src/main/java/org/elasticsearch/rest/RestController.java:
```diff
@@ -104,25 +107,32 @@ public RestController(Set<RestHeaderDefinition> headersToCopy, UnaryOperator<Res
         assert (handler instanceof DeprecationRestHandler) == false;
 
-        registerHandler(method, path, new DeprecationRestHandler(handler, deprecationMessage, deprecationLogger));
+        if (version.major >= RestApiCompatibleVersion.currentVersion().major) {
```
> not sure if there is a possibility to `version.major` be ever greater 

d3e008111fac2ca58b2ea3d94295b8142a2bfb1b

joegallo replied:
78f4726 covers the NPE, `restApiVersion` will always be there

joegallo committed:
Refactor Route so that restApiVersion is never null

joegallo commented on server/src/main/java/org/elasticsearch/rest/RestHandler.java:
```diff
@@ -96,24 +78,84 @@ default boolean allowSystemIndexAccessByDefault() {
+
+        // bog standard
+        public Route (Method method, String path) {
+            this(method, path, null, null, null, null);
```
78f47266ea476d92ec5cb249eb2e53616dd87434 ✅

jakelandis commented:
@joegallo - can you provide a simple example of before and after in the primary comment / commit message ?

jakelandis commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/rest/action/user/RestChangePasswordAction.java:
```diff
@@ -46,21 +47,16 @@ public RestChangePasswordAction(Settings settings, SecurityContext securityConte
-
-    @Override
-    public List<ReplacedRoute> replacedRoutes() {
         // TODO: remove deprecated endpoint in 8.0.0
```
Does this change satisfy this TODO ?

joegallo replied:
5bee773347657480feb8e1cec58776d68d4943b3, 👍, yes, I agree that it does.

joegallo committed:
Remove TODOs

joegallo committed:
Rewrite this test so it's exercising the intended behavior

elasticmachine commented:
Pinging @elastic/es-core-infra (Team:Core/Infra)

joegallo commented:
@jaymode @rjernst can you take a look at this one? 

It's not hugely different than the last time you saw it, @jaymode, but it's much less WIP-y now and (hopefully) closer to mergeable.

pgomulka commented on server/src/main/java/org/elasticsearch/rest/RestController.java:
```diff
@@ -104,25 +107,34 @@ public RestController(Set<RestHeaderDefinition> headersToCopy, UnaryOperator<Res
         assert (handler instanceof DeprecationRestHandler) == false;
-
-        registerHandler(method, path, new DeprecationRestHandler(handler, deprecationMessage, deprecationLogger));
+        if (version.major == RestApiVersion.current().major) {
```
would it be better we actually use `version == RestApiVersion.current()` instead of `.major` ? Those are enums, so comparison should look good I think.
THe same for `version.major < RestApiVersion.current().major` Are we expecting a version to be ever different then minimumSupported here (N-1) ? how about we just use `version == RestApiVersion.minimuSupported`

With that we would prevent registering N-2 handlers after the version bump was performed, but N-2 compatible handlers were not removed yet.

pgomulka commented on server/src/main/java/org/elasticsearch/rest/RestController.java:
```diff
@@ -104,25 +107,34 @@ public RestController(Set<RestHeaderDefinition> headersToCopy, UnaryOperator<Res
         assert (handler instanceof DeprecationRestHandler) == false;
-
-        registerHandler(method, path, new DeprecationRestHandler(handler, deprecationMessage, deprecationLogger));
+        if (version.major == RestApiVersion.current().major) {
```
would it be better we actually use version == RestApiVersion.current() instead of .major ? Those are enums, so comparison should look good I think.
THe same for version.major < RestApiVersion.current().major Are we expecting a version to be ever different then minimumSupported here (N-1) ? how about we just use version == RestApiVersion.minimuSupported

With that we would prevent registering N-2 handlers after the version bump was performed, but N-2 compatible handlers were not removed yet.

Also not sure the second comment is right. `            // e.g. it was marked as deprecated in 8.x, and we're currently running 7.x
`
shouldn't it be ?
`            // e.g. it was marked as deprecated in 7.x, and we're currently running 8.x`

joegallo replied:
We chatted offline and 0149fe680d410a81af2c9cd2bcbdbd2b0bb4d27b should cover this.

pgomulka reviewed (commented):
I left 2 comments. I feel it would be better not to include the nested classes imports in this PR. It is big itself, and given we also backport it, then it would be safer to keep it as self contained as possible
Other then those 2 comments It looks good to me

joegallo committed:
Twiddle imports

prefer static reference to RestRequest.Method members for Route
definitions in RestActions.

joegallo committed:
Twiddle imports

generally prefer qualified RestRequest.Method everywhere besides when
defining Routes in RestActions

joegallo committed:
Compare enums and fix a comment

joegallo commented:
@pgomulka 7bb0919 should address your comment w.r.t. imports

pgomulka reviewed (approved):
LGTM, thank you!

joegallo committed:
Merge remote-tracking branch 'upstream/master' into route-refactoring

joegallo committed:
Fix some docstrings

jaymode commented on server/src/main/java/org/elasticsearch/rest/DeprecationRestHandler.java:
```diff
@@ -34,10 +35,16 @@
      * @throws IllegalArgumentException if {@code deprecationMessage} is not a valid header
      */
-    public DeprecationRestHandler(RestHandler handler, String deprecationMessage, DeprecationLogger deprecationLogger) {
+    public DeprecationRestHandler(RestHandler handler, String deprecationMessage, DeprecationLogger deprecationLogger,
```
can you update the javadocs?

joegallo replied:
Argh, I fall on my sword. Good catch.

joegallo replied:
9833cbf

jaymode commented on server/src/main/java/org/elasticsearch/rest/DeprecationRestHandler.java:
```diff
@@ -34,10 +35,16 @@
+        this.compatibleVersionWarning = compatibleVersionWarning;
+    }
+
+    public DeprecationRestHandler(RestHandler handler, String deprecationMessage, DeprecationLogger deprecationLogger) {
```
can you add javadocs?

joegallo replied:
9833cbf

jaymode commented on server/src/main/java/org/elasticsearch/rest/MethodHandlers.java:
```diff
@@ -22,26 +22,21 @@
-            methodHandlers.computeIfAbsent(method, k -> new HashMap<>())
-                .put(handler.compatibleWithVersion(), handler);
-        }
+        this.methodHandlers = new HashMap<>(4);
```
can you add a comment about the choice of `4` for the initial map size?

joegallo replied:
Good call, 9afd325cd756b318137f51804badf52cc8e962db.

jaymode commented on server/src/main/java/org/elasticsearch/rest/MethodHandlers.java:
```diff
@@ -22,26 +22,21 @@
-            }
+    MethodHandlers addMethod(RestRequest.Method method, RestApiVersion version, RestHandler handler) {
+        RestHandler existing = methodHandlers
+            .computeIfAbsent(method, k -> new HashMap<>())
```
I wonder about the sizing here. For the majority of APIs, I think we would have at most two items in this map. So we would want the ceiling of 2 / 0.75, which is 3?

joegallo replied:
I went in a slightly different direction, but yes, 100% and great catch that this was using the default initialCapacity, yikes. 9afd325cd756b318137f51804badf52cc8e962db

jaymode reviewed (approved):
I left a few minor comments, otherwise LGTM.

joegallo committed:
Adjust the initialCapacity and loadFactor for these HashMaps

and add comments explaining their size

joegallo committed:
Clean up DeprecationRestHandler

Collapse two constructors into one, document the new parameter that it
grew, and add a test that exercises the new parameter.

joegallo commented:
Thank you @jakelandis, @jaymode, and @pgomulka for working with me on this one. I'm really happy with where it ended up!

Now to resurrect the `_xpack/*` APIs under this new mechanism. 😄

joegallo committed:
Merge remote-tracking branch 'upstream/master' into route-refactoring

joegallo committed:
Whitespace
